### PR TITLE
Introduce @JsonIgnoreUnknownKeys annotation

### LINKED ISSUE #1420: Ignore unknown keys per class / via annotation
**What is your use-case and why do you need this feature?**

I know [keys can be ignored "globally"](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/json.md#ignoring-unknown-keys) on the JSON builder scope. But for safely, and to better document which classes have been implemented completely and which not, I'd prefer to ignore unknown keys per class only.

**Describe the solution you'd like**

I'd prefer an annotation-based solution, similar to Jackson's `@JsonIgnoreProperties(ignoreUnknown = true)` annotation.

### LINKED ISSUE #2637: Improve failure message on encountering unknown key in JSON
Right now, the failure is the following:
```
Exception in thread "main" kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 20: Encountered an unknown key 'version' at path: $.name
Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
JSON input: {"name":"unknown", "version": 2.0}
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
```

Maybe it is worth making it more explicit and straightforward, for example:
```
Exception in thread "main" kotlinx.serialization.json.internal.JsonUnknownKeyException: Encountered an unknown key 'version' at path: $.name
Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
JSON input: {"name":"unknown", "version": 2.0}
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
```

Note that the actual exception can still be internal

### LINKED ISSUE #2869: Wrong path reported by json parser for unknown key
**Describe the bug**
When parsing json with unknown key, the path in the error message will only be correct if the unknown key is the first key in the object. If it is not the first key, the reported path will include the previous key at the same level as the unknown key.

**To Reproduce**
```
package ...

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class UnknownKeyParsingTest {
    @Serializable
    data class MyClass(val field: String)

    @Test
    fun testUnknownKey() {
        assertThrows<SerializationException> {
            Json.decodeFromString<MyClass>("""{ "unknownField": 123 }""")
        }.also {
            assertEquals(
                """
                    Unexpected JSON token at offset 3: Encountered an unknown key 'unknownField' at path: $
                    Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
                    JSON input: { "unknownField": 123 }
                """.trimIndent(),
                it.message
            )
        }
    }

    @Test
    fun testUnknownKeyAsSecondKey() {
        assertThrows<SerializationException> {
            Json.decodeFromString<MyClass>("""{ "field": "value", "unknownField": 123 }""")
        }.also {
            assertEquals(
                """
                    Unexpected JSON token at offset 21: Encountered an unknown key 'unknownField' at path: $.field
                    Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.
                    JSON input: { "field": "value", "unknownField": 123 }
                """.trimIndent(),
                it.message
            )
        }
    }
}
```
**Expected behavior**
The expected behaviour is that the reported path is for the object that contains the unknown key, not for a sibling key.

**Environment**
 - Kotlin version: 2.0.21
 - Library version: 1.7.3
 - Kotlin platforms: JVM

### PR DESCRIPTION
**Pull request by sandwwraith**
for more fine-grained control over JsonBuilder.ignoreUnknownKeys setting.

Fixes #1420

Also improve error message and path handling when 'Unknown key' exception is thrown

Fixes #2869
Fixes #2637

### CODE REVIEW

File: formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt [hunk 1]
```diff
@@ -74,3 +74,36 @@ public annotation class JsonNames(vararg val names: String)
 @Target(AnnotationTarget.CLASS)
 @ExperimentalSerializationApi
 public annotation class JsonClassDiscriminator(val discriminator: String)
+
+
+/**
+ * Specifies whether encounters of unknown properties (i.e., properties not declared in the class) in the input JSON
+ * should be ignored instead of throwing [SerializationException].
+ *
+ * With this annotation, it is possible to allow unknown properties for specific classes, while
```
Comment (fzhinkin) [inline formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt:83]: Annotation's effect does not propagate to annotated class' fields:

```kotlin
@JsonIgnoreUnknownKeys @Serializable
class Outer(val inner: Inner)

@Serializable
class Inner(val x: Int)

...
// {"untitled":"unknown", "inner": {"x": 42}} // OK
// {"inner": {"x": 42, "untitled": "unknown"}} // FAIL
```

Was it a deliberate choice?
Personally, I think the current behavior is fine, but it might be worth to stress that `for specific classes` means that the annotation has a shallow effect.

File: guide/example/example-json-05.kt [hunk 2]
```diff
@@ -4,16 +4,13 @@ package example.exampleJson05
 import kotlinx.serialization.*
 import kotlinx.serialization.json.*
 
-val format = Json { encodeDefaults = true }
```
Comment (fzhinkin) [inline guide/example/example-json-05.kt:7]: git's capabilities of tracking renames are quite limited and it seems like it can't handle such a chain of renames.

I see two possible solutions to keep files' git history a bit more clear:
- commit all renames separately (wow, 28 commits 😓 )
- put the new example into `example-json-32.kt` instead of 04.

WDYT?

File: formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt (hunk 1 above)
Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt:83]: Yes, it is a deliberate choice (also we don't have technical means for this). 
I'll change 'specific' to 'annotated'.

File: guide/example/example-json-05.kt (hunk 2 above)
Comment (sandwwraith) [reply guide/example/example-json-05.kt:7]: I'm afraid Knit functionality as a tool is quite limited in this place — it will automatically re-number all examples in a sequence (https://github.com/Kotlin/kotlinx-knit?tab=readme-ov-file#example-files)

File: docs/json.md [hunk 3]
```diff
@@ -164,6 +165,44 @@ Project(name=kotlinx.serialization)
...
+    println(Json.decodeFromString<Outer>("""{"a":1,"inner":{"x":"value","unknownKey":"unknownValue"}}"""))
+    println()
+    // 2
+    println(Json.decodeFromString<Outer>("""{"a":1,"inner":{"x":"value"}, "b":2}"""))
+}
+```
+
+> You can get the full code [here](../guide/example/example-json-04.kt).
+
+Line (1) decodes successfully despite "unknownKey" in `Inner`, because annotation is present on the class.
```
[the remark this answers is not in the document]
Comment (sandwwraith) [reply docs/json.md:194]: To illustrate 'shallow' effect? I think I can just move the annotation

File: formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt [hunk 4]
```diff
@@ -74,3 +74,36 @@ public annotation class JsonNames(vararg val names: String)
...
 @ExperimentalSerializationApi
 public annotation class JsonClassDiscriminator(val discriminator: String)
+
+
+/**
+ * Specifies whether encounters of unknown properties (i.e., properties not declared in the class) in the input JSON
+ * should be ignored instead of throwing [SerializationException].
+ *
+ * With this annotation, it is possible to allow unknown properties for annotated classes, while
+ * general [Json.decodeFromString] would still reject them for everything else.
```
Comment (shanshin) [inline formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt:84]: Perhaps I'm exaggerating, but won't such a description give the user the impression that this annotation only works with `decodeFromString`, but not with `decodeFromJsonElement` or `decodeFromStream`?

**Review by shanshin — APPROVED** [review]

File: docs/json.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fzhinkin) [reply docs/json.md]: Yeap, I was talking about the 'shallow' effect.

**Review by fzhinkin — APPROVED** [review]

File: formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt (hunk 4 above)
Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/JsonAnnotations.kt:84]: Not sure about that. Maybe rephrase as `general decoding methods, such as [Json.decodeFromString] and others`?
